File: common/scala/src/main/scala/org/aiddl/common/scala/planning/state_variable/data/DomainTransitionGraphCreator.scala
package org.aiddl.common.scala.planning.state_variable.data

import scala.collection.mutable.{HashMap, HashSet}
import scala.collection.{immutable, mutable}
import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.representation.*
import org.aiddl.common.scala.math.graph.Terms.*
import org.aiddl.common.scala.planning.PlanningTerm.*
import org.aiddl.common.scala.planning.state_variable.heuristic.CausalGraphHeuristic.Unknown
import org.aiddl.core.scala.representation.conversion.given_Conversion_Term_KeyVal

import scala.language.implicitConversions


class DomainTransitionGraphCreator(problem: CollectionTerm) extends Function {

  def apply( os: Term ): Term = {
    val domains = new mutable.HashMap[Term, mutable.HashSet[Term]]

    problem(InitialState).asCol.foreach( s => {
      domains.getOrElseUpdate(s.key, new mutable.HashSet).add(s.value)
    })

    os.asCol.foreach( o => {
        o(Preconditions).asCol.foreach( p => { domains.getOrElseUpdate(p.key, new mutable.HashSet).add(p.value) })
        o(Effects).asCol.foreach( p => { domains.getOrElseUpdate(p.key, new mutable.HashSet).add(p.value) })
    })

    SetTerm(domains.keys.map( v => {
        KeyVal(v, dtg(v, domains(v), os.asCol))
    } ).toSet)
  }

  def dtg(v: Term, domain: mutable.HashSet[Term], operators: CollectionTerm ): Term = {
    var edges = new immutable.HashSet[Term]
    val labels = new HashMap[Term, HashSet[Term]]

    operators.withFilter( o => o(Effects).asCol.containsKey(v)
      && o(Preconditions).getOrElse(v, Unknown) != o(Effects)(v)  )
      .foreach( o => {
        val from = o(Preconditions).getOrElse(v, Unknown)
        val edge = Tuple(from, o(Effects)(v))
        val cond = SetTerm(o(Preconditions).asCol.filter( p => p.key != v || p.value != from ).toSet)

        labels.getOrElseUpdate(edge, new HashSet).add(cond)
        edges = edges + edge

        if ( from == Unknown ) { // Add edge from each value of the domain other than target
          domain.foreach( from => {
            if ( from != o(Effects)(v) ) {
              val edge = Tuple(from, o(Effects)(v))
              val cond = SetTerm(o(Preconditions).asCol.filter( p => p.key != v || p.value != from ).toSet)
              labels.getOrElseUpdate(edge, new HashSet).add(cond)
              edges = edges + edge
            }
          })
        }
      } )
    Tuple(
      KeyVal(Nodes, SetTerm(domain.toSet)),
      KeyVal(Edges, SetTerm(edges)),
      KeyVal(Labels, SetTerm(labels.map( (k, c) => KeyVal(k, SetTerm(c.toSet)) ).toSet))
    )
  }
}

File: common/java/src/main/java/org/aiddl/common/java/planning/state_variable/data/DomainTransitionGraphCreator.java
package org.aiddl.common.java.planning.state_variable.data;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.aiddl.common.java.math.graph.GraphTerm;
import org.aiddl.common.java.planning.state_variable.heuristic.CausalGraphHeuristic;
import org.aiddl.common.java.planning.state_variable.Operator;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.Term;
import org.aiddl.core.java.tools.LockableList;
import org.aiddl.core.java.tools.LockableSet;

public class DomainTransitionGraphCreator implements Function {

	@Override
	public Term apply(Term O) {
		Map<Term, Set<Term>> domains = new HashMap<>();
		Set<Term> variables = new HashSet<>();
		for ( Term o_term : O.asCollection() ) {
			Operator o = new Operator(o_term.asTuple());
			for ( Term e : o.getEffects() ) {
				domains.putIfAbsent(e.getKey(), new HashSet<>());
				domains.get(e.getKey()).add(e.getValue());
				variables.add( e.getKey() );
			}
			for ( Term p : o.getPreconditions() ) {
				domains.putIfAbsent(p.getKey(), new HashSet<>());
				domains.get(p.getKey()).add(p.getValue());
				variables.add( p.getKey() );
			}
		}
		LockableSet DTGs = new LockableSet();
		
		for ( Term variable : variables ) {
			DTGs.add(Term.keyVal(variable, compute(O, variable, domains.get(variable))));
		}
		
		return Term.set(DTGs);
	}
	

	public Term compute(Term O, Term variable, Set<Term> domain ) {
		LockableSet V = new LockableSet();
		LockableList E = new LockableList();
		LockableSet L = new LockableSet();
		
		for ( Term o_term : O.asCollection() ) {
			Operator o = new Operator(o_term.asTuple());
			
			Term preVal = o.getPreconditions().getOrDefault(variable, CausalGraphHeuristic.Unknown);
			Term effVal = o.getEffects().get(variable); 

			Set<Term> from_values = null;

			from_values = new HashSet<>();
			from_values.add(preVal);
		
			for ( Term from_value : from_values ) {
				if ( effVal != null && !from_value.equals(effVal) ) {
					Term edge = Term.tuple(from_value, effVal);
//					LockableSet edge = new LockableSet();
//					edge.add(Term.keyVal(GraphTerm.Edge, ));
					
//					System.out.println(variable + " -> " + from_value);
					LockableSet conditions = new LockableSet();
					for ( Term p : o.getPreconditions() ) {
						if ( !(p.getKey().equals(variable) && p.getValue().equals(from_value)) ) {
							conditions.add(p);
						}
					}
//					System.out.println(Logger.prettyPrint(Term.set(conditions), 0));
//					Term e = Term.tuple(from_value, effVal); 
					E.add(edge);
//					edge.add(Term.keyVal(GraphTerm.Label, )); //o.getPreconditions().asCollection().remove(Term.keyVal(variable, from_value))));
					L.add(Term.keyVal(edge, Term.set(conditions)));
//					System.out.println(Term.set(edge));

					V.add(from_value);
					V.add(effVal);
				}
			}
		}
		return Term.tuple(
				Term.keyVal(GraphTerm.Nodes, Term.set(V)),
				Term.keyVal(GraphTerm.Edges, Term.set(E)),
				Term.keyVal(GraphTerm.Labels, Term.set(L))
				);
	}
}
